### PR TITLE
improve python3 support

### PR DESCRIPTION
**Pull request by gijzelaerr**

### CODE REVIEW

**Comment by mr-c** [comment]
"This branch is out-of-date with the base branch" Should I hit the 'update branch' button, or do you want to do that locally?

File: cwltool/expression.py [hunk 1]
```diff
@@ -24,7 +25,7 @@ def jshead(engineConfig, rootvars):
 segment_re = re.compile(segments, flags=re.UNICODE)
 param_re = re.compile(r"\((%s)%s*\)$" % (seg_symbol, segments), flags=re.UNICODE)
 
-JSON = Union[Dict[Any, Any], List[Any], Text, int, long, float, bool, None]
+JSON = Union[Dict[Any, Any], List[Any], Text, int, int, float, bool, None]
```
Comment (mr-c) [inline cwltool/expression.py]: hmm...

File: cwltool/sandboxjs.py [hunk 2]
```diff
@@ -18,7 +17,7 @@ class JavascriptException(Exception):
 
 _logger = logging.getLogger("cwltool")
 
-JSON = Union[Dict[Text, Any], List[Any], Text, int, long, float, bool, None]
+JSON = Union[Dict[Text, Any], List[Any], Text, int, int, float, bool, None]
```
Comment (mr-c) [inline cwltool/sandboxjs.py]: Oh, didn't know we duplicated this with https://github.com/common-workflow-language/cwltool/pull/287/files#diff-cb6110fc9c5f825722db0f9c57cb1335R28

File: cwltool/sandboxjs.py [hunk 3]
```diff
@@ -1,12 +1,11 @@
-import cStringIO
```
Comment (mr-c) [inline cwltool/sandboxjs.py]: This is needed on line 107

File: cwltool/expression.py [hunk 4]
```diff
@@ -126,7 +127,7 @@ def next_seg(remain, obj):  # type: (Text, Any) -> Any
             try:
                 key = int(m.group(0)[1:-1])
             except ValueError as v:
-                raise WorkflowException(unicode(v))
+                raise WorkflowException(u(v))
```
Comment (mr-c) [inline cwltool/expression.py]:
```
cwltool/expression.py: note: In function "next_seg":
cwltool/expression.py:130: error: Argument 1 to "u" has incompatible type "ValueError"; expected "str"
```

File: cwltool/draft2tool.py [hunk 5]
```diff
@@ -406,7 +407,7 @@ def collect_output_ports(self, ports, builder, outdir, compute_checksum=True):
                                 % shortname(port["id"]), exc_info=True)
                             raise WorkflowException(
                                 u"Error collecting output for parameter '%s':\n%s"
-                                % (shortname(port["id"]), indent(unicode(e))))
+                                % (shortname(port["id"]), indent(u(e))))
```
Comment (mr-c) [inline cwltool/draft2tool.py]:
```
cwltool/draft2tool.py: note: In member "collect_output_ports" of class "CommandLineTool":
cwltool/draft2tool.py:410: error: Argument 1 to "u" has incompatible type "Exception"; expected "str"
```

File: cwltool/sandboxjs.py (hunk 3 above)
Comment (gijzelaerr) [inline cwltool/sandboxjs.py]: ah yes. clearly this is not covered by the test suite :)

File: cwltool/expression.py (hunk 4 above)
Comment (gijzelaerr) [inline cwltool/expression.py]: where did you find these errors?

File: cwltool/sandboxjs.py (hunk 3 above)
Comment (gijzelaerr) [inline cwltool/sandboxjs.py]: well, it is not required to run the code, only to validate the types. But there are many more type violations due to missing imports in the code... so what should we do?

File: cwltool/expression.py (hunk 4 above)
Comment (mr-c) [inline cwltool/expression.py]: `make mypy` and https://travis-ci.org/common-workflow-language/cwltool/jobs/199538581#L185

File: cwltool/sandboxjs.py (hunk 3 above)
Comment (mr-c) [inline cwltool/sandboxjs.py]: The lines I commented on are the only type violations I'm aware of.

Comment (gijzelaerr) [inline cwltool/sandboxjs.py]: pycharm finds a lot, but that is mostly things like `typing.List` and `typing.Union`, which are probably imported by default by mypy.

Comment (mr-c) [inline cwltool/sandboxjs.py]: Oh, those should be added in, yes.

Comment (gijzelaerr) [inline cwltool/sandboxjs.py]: it looks impossible to make that type annotation work properly with Python3, the stringio stuff works quite different now.

File: cwltool/sandboxjs.py [hunk 6]
```diff
@@ -101,9 +100,9 @@ def term():
...
     tm.start()
 
-    stdin_buf = StringIO(json.dumps(fn) + "\n")
-    stdout_buf = StringIO()
-    stderr_buf = StringIO()
+    stdin_buf = BytesIO(json.dumps(fn) + "\n")
+    stdout_buf = BytesIO()
+    stderr_buf = BytesIO()
 
     completed = []  # type: List[Union[cStringIO.InputType, cStringIO.OutputType]]
```
Comment (mr-c) [inline cwltool/sandboxjs.py]: I'd try `completed = []  # type: List[BytesIO]` here ...

**Comment by mr-c** [comment]
and add 3.4 & 3.6 to `.travis.yml' to turn those tests on

File: cwltool/sandboxjs.py (hunk 6 above)
Comment (gijzelaerr) [inline cwltool/sandboxjs.py]: yep, but that is python3 only.

Comment (mr-c) [inline cwltool/sandboxjs.py]: I updated `schema_salad` and `cwltool` for `mypy` 0.470 (the project was renamed from `mypy-lang`). With that in place `type: List[BytesIO]` works for me. If you'll merge in from master then I think we'll be done with this PR.

**Pull request merged**